Migration guide (#1687)
typelevel/cats-effect: docs/migration-guide.md

---
id: migration-guide
title: Migration guide
---

## Summary

> Here is a general view of the steps you should take to migrate your application to Cats Effect 3:

Reviewer: "a general view" -> "an overview"?
```suggestion
Here is an overview of the steps you should take to migrate your application to Cats Effect 3:
```

1. [Make sure your dependencies have upgraded](#make-sure-your-dependencies-have-upgraded)<!-- don't remove this comment - this ensures the vscode extension doesn't make this a ToC -->
1. [Run the Scalafix migration](#run-the-scalafix-migration) (optional)
1. [Upgrade dependencies and Cats Effect itself](#upgrade-dependencies)
1. [Fix remaining compilation issues](#fix-remaining-compilation-issues)
1. [Test your application](#test-your-application).

### Before you begin: this isn't a "quick start" guide

This guide is meant for existing users of Cats Effect 2 who want to upgrade their applications
to 3.0.0.

> If you haven't used Cats Effect before and want to give it a try,
> please follow the [getting started guide](./getting-started.md) instead!

### 🤔 Need help?

If any point of the migration turns out to be difficult and you feel like you need help, feel free to [explain your problem on Gitter](https://gitter.im/typelevel/cats-effect) and we will do our best to assist you.

> If you spot a mistake in the guide or the library itself, you can [report an issue on GitHub](https://github.com/typelevel/cats-effect/issues/new).

Reviewer: Maybe also nudge towards a PR?

Author: adding in f86f4ddd5457fe14ecfcbd019c3a60bfb3f91949

To find out more, see [its docs](./std/dispatcher.md).

For example, given a bogus library interface like this:

```scala mdoc
trait LibraryInterface {
  def addListener(f: String => Unit): Unit
  def run(): Unit // starts a task in the background
  def shutdown(): Unit
}
```

In CE2 you could run an effect in `addListener` using `Effect` like this:

```scala
// CE2
def addListenerF[F[_]: Effect](
  interface: LibraryInterface,
  listener: String => F[Unit]
): F[Unit] =
  Sync[F].delay {
    interface.addListener(str => listener(str).toIO.unsafeRunSync())
    interface.run()
  } *> Async[F].never.guarantee(Sync[F].delay(interface.shutdown()))
```

In CE3, you would use `Dispatcher`:

```scala mdoc
import cats.effect.Async
import cats.effect.Sync
import cats.effect.std.Dispatcher
import cats.effect.syntax.all._

// CE3
def addListenerF[F[_]: Async](
  interface: LibraryInterface,
  listener: String => F[Unit]
): F[Unit] =
  Dispatcher[F].use { dispatcher =>
    Sync[F].delay {
      interface.addListener(str => dispatcher.unsafeRunSync(listener(str)))
      interface.run()
    } *> Async[F].never[Unit].guarantee(Sync[F].delay(interface.shutdown()))
  }

> ```

Reviewer: Another dispatcher example illustrating the case where it isn't immediately `.use`d might be helpful to illustrate the lifecycle mentioned above. My concern is that the current example easily becomes a footgun if someone were to modify it for a case where they didn't immediately block with `.never`.

```scala
trait Consumer[A] {
  def onNext(a: A): Unit
}
```

``` scala
// CE2
def consumer[F[_]: Effect, A](handler: A => F[Unit]): Consumer[A] =
  new Consumer[A] {
    def onNext(a: A) = handler(a).toIO.unsafeRunSync()
  }
```

```scala
// CE3
def consumer[F[_]: Async, A](handler: A => F[Unit]): Resource[F, Consumer[A]] =
  Dispatcher[F] map { dispatcher =>
    new Consumer[A] {
      def onNext(a: A) = dispatcher.unsafeRunSync(handler(a))
    }
  }
```

```scala
// CE3 alternative
def consumer[F[_], A](dispatcher: Dispatcher[F], handler: A => F[Unit]): Consumer[A] =
  new Consumer[A] {
    def onNext(a: A) = dispatcher.unsafeRunSync(handler(a))
  }
```

Author: actually, I'll just use that. It's a much simpler example.

final case class Succeeded[F[_], E, A](fa: F[A]) extends Outcome[F, E, A]
final case class Errored[F[_], E, A](e: E) extends Outcome[F, E, A]
final case class Canceled[F[_], E, A]() extends Outcome[F, E, A]
```

If we ignore the various differences in type parameter variance (which were mostly added for better Scala 3 support),
the elephant in the room is `Succeeded` (the new `Completed`) - it has an `F[A]` field, which will contain the value the effect produced in the successful case. This means methods like `bracketCase` / `Resource.allocateCase` can use this result when cleaning up the resource.

If you're simply migrating code that was using these methods, just renaming to the appropriate new names (and possibly fixing some pattern matches)
should get you to a compiling state. For more information about Outcome, see [`Spawn` docs](./typeclasses/spawn.md).

#### `Fiber`

Mostly unchanged, `Fiber` still has a `cancel` method, and a `join` method:

```scala
trait Fiber[F[_], E, A] {
  def cancel: F[Unit]
  def join: F[Outcome[F, E, A]]
}
```

However, there are still some differences here: first of all, `join` doesn't just return `F[A]`, but the whole `Outcome` of it.

> This is also why `Fiber` got the extra type parameter `A`.

Reviewer 2: Maybe E?

Author: true lol, I made a huge mistake here. Thanks!
